### ISSUE TITLE
Have you verified whether the abort signal is actually functioning in the edge runtime?

### ISSUE THREAD

**Issue by wsxiaoys** [vercel-labs/ai#90]
Based on my experiments so far, it appears that the AbortController doesn't function properly on Vercel's hosted edge runtime. This observation aligns with the issue discussed in detail on the following GitHub issue: https://github.com/vercel/next.js/issues/50364.

**Comment by jensen** [comment]
I am coming to the same conclusion. 

About a month ago I was asked to add cancellation to a product I am working on for a client. I spent some time attempting to get this working with edge functions, but could not receive the cancel callback on the server. I decided to see if I could get it working at all, and was able to implement a version using Deno since it has a similar runtime API.

At this point I sent a support request to Vercel and about two weeks later it was confirmed the upstream provider, Cloudflare I assume, did not support the abort signal and was aware of this limitation.

I ended up moving our Open AI endpoints to a Fastify server running Node 20 on AWS an ECS cluster. This setup has allowed me to handle the cancellations properly. I am able to cancel my request to OpenAI which saves me tokens and it writes whatever tokens that were received to the db with a "cancel" type.

I was very excited to see the release of this library today, particularly this documentation https://sdk.vercel.ai/docs/concepts/backpressure-and-cancellation.

I did not expect the edge function to receive the AbortController signal, even though the examples seem to use this as the cancellation mechanism. The example of using the `pull` callback in a ReadableStream was something I hadn't tried previously so I decided to apply the approach to our codebase.

I found that the AIStream doesn't use the `pull` callback, but instead has a TransformStream. I looked for `pull` and found it in the Hugging Face example. I ended up converting my request.body to an iterator using a similar approach. This also doesn't seem to handle the cancellation. 

I am using the pages directory in our app, I am assuming that the examples only use the new app directory to highlight it, rather than it being a requirement.

**Comment by wsxiaoys** [comment]
> I moved our Open AI streaming endpoints to a Fastify server running Node 20 on AWS an ECS cluster. This setup has allowed me to handle the cancellations properly. I can cancel my request to OpenAI, which saves me tokens and writes whatever tokens were received to the db with a "cancel" type.

@jensen were you able to confirm this behavior? Based on my experiments, it seems that sending a cancel signal to OpenAI does not actually reduce the token usage of the request...

**Comment by jensen** [comment]
There are a lot of moving pieces, but I have successfully cancelled the request using both Deno and Fastify with Node 20. 

I have my own Open AI API account that has no traffic. I use this to confirm a long completion that is cancelled uses the number of tokens that matches what I calculate using tiktoken vs what it would have if it finished. My prompt would be "Write a blog post about React". When I cancel it after a few sentences, the usage on the Open AI dashboard matches.

**Comment by wsxiaoys** [comment]
> There are a lot of moving pieces, but I have successfully cancelled the request using both Deno and Fastify with Node 20.
> 
> I have my own Open AI API account that has no traffic. I use this to confirm a long completion that is cancelled uses the number of tokens that matches what I calculate using tiktoken vs what it would have if it finished. My prompt would be "Write a blog post about React". When I cancel it after a few sentences, the usage on the Open AI dashboard matches.

It's good to know that cancellation is indeed possible with OpenAI's API. Now, the next steps lie in vercel / nextjs side...

**Comment by jensen** [comment]
It looks like this was released today. https://github.com/vercel-labs/ai-chatbot

It has a stop-generating button https://chat.vercel.ai/ which makes sense since the SDK has a stop API through the hooks. I guess my next step is to clone this, set up some logging and deploy it to Vercel to double-check how it is behaving on the server. Perhaps the development environment isn't a good one to test this in, it hasn't been in the past.

**Comment by wsxiaoys** [comment]
> It looks like this was released today. https://github.com/vercel-labs/ai-chatbot
> 
> It has a stop-generating button https://chat.vercel.ai/ which makes sense since the SDK has a stop API through the hooks. I guess my next step is to clone this, set up some logging and deploy it to Vercel to double-check how it is behaving on the server. Perhaps the development environment isn't a good one to test this in, it hasn't been in the past.

Thank you for taking the time to verify this. However, based on the source code, I predict that it won't work because the `stop` button simply calls the `AbortController.abort()` function.

I'm eagerly anticipating the result.

**Comment by enricoros** [comment]
Seems like it stops the new tokens to be displayed on the screen, but the token production still goes on between the Vercel Edge and OpenAI, so you'll be charged for the full amount and incur rate limiting while many generations are still going on in the background.
High priority fix, imo.

**Comment by jensen** [comment]
I was able to spend some time testing this tonight. I deployed a version of the ai-chatbot application with some additional logging. I am pretty confident that the cancellation is working as intended when using the edge runtime. 

This is excellent news. 

I haven't done as much testing as I want to for our production application, but my next step will be to test my `pull` based streaming changes using our staging environment. 

I likely won't get to this in the next few days since we have already shipped our cancellation feature using AWS. I will still want to move these endpoints back to Vercel in July.

**Comment by wsxiaoys** [comment]
Seems related:

https://github.com/vercel/edge-runtime/pull/396
https://github.com/vercel/next.js/pull/51330

**Comment by jridgewell** [comment]
Hi! Yes, I'm working on getting proper streaming cancellation and back-pressure into Next.js. If you're using Next as your dev/production server, it's not currently possible to end the stream. Once https://github.com/vercel/next.js/pull/51330 is merged and released, this should be fixed.

**Comment by jridgewell** [comment]
Hi all! We've merged https://github.com/vercel/next.js/pull/51594, which implements cancellation **only**. We'll work on getting back-pressure support after verifying it's impact on Next's general streaming performance (Next is mainly for streaming React components, and we need to make sure that's not taking a hit). I don't think it's going to be an issue, but just need some time to verify.

**Comment by enricoros** [comment]
@jridgewell Anything special devs need to do for using cancellation? And what is back-pressure and what do we need to accomodate it?

I work on one of the popular opensource GPT UIs (https://github.com/enricoros/big-agi) and I'm sure devs like us appreciate your fix.

**Comment by jridgewell** [comment]
> Anything special devs need to do for using cancellation?

The Next.js team hasn't released a new version yet (I'll ping them to see if they can do a canary release), but once they have users just need to `npm install next@latest` (or `next@canary`).

> And what is back-pressure and what do we need to accomodate it?

Back-pressure is explained in https://sdk.vercel.ai/docs/concepts/backpressure-and-cancellation. Essentially, it's the ability for the server pause the stream because the client doesn't need more information yet. Next.js hasn't added support for it yet, but when they do, everyone will need to update their `next` dependency again.

**Comment by enricoros** [comment]
@jridgewell thanks for the explanation. Will try out the canary when available. I'm glad it doesn't require code changes (maybe some exception handling?) I was trying with AbortControllers and exceptions everywhere, but nothing worked for me wrt cancellations.

**Comment by jridgewell** [comment]
[v13.4.8-canary.0](https://github.com/vercel/next.js/releases/tag/v13.4.8-canary.0 ) just got released. If you update your project dependency, your dev server will support cancellation, and when you deploy that change your prod server should get it too.

**Comment by enricoros** [comment]
@jridgewell I tried canary.0 and .1, but somehow is not working for me. Maybe it's a different issue.

When I close the socket to the server (physically closing the Chrome window of the edge function caller), this is what I see in the log of the dev server (`next dev`):
![image](https://github.com/vercel-labs/ai/assets/32999/0fdf6645-04ba-4825-8266-146f5905415d)

And this is the code that prints the streaming events (the error is printed by the edge server):

![image](https://github.com/vercel-labs/ai/assets/32999/3cafd2fc-259c-4834-985e-db3d1dd39e4f)

This is within the ReadableStream.
```
return new ReadableStream({
    start: (controller) => {
        ...loop above...
    },
});
```

I must be doing something wrong.

**Comment by jvandenaardweg** [comment]
Shouldn't you catch the AbortError on the server and expect it to happen by returning null or something? Like the client hooks do: https://github.com/vercel-labs/ai/blob/main/packages/core/react/use-completion.ts#L179

**Comment by enricoros** [comment]
Tried catching the AbortError but doesn't catch anything. Agree, the token usage keeps skyrocketing, a sign that the request to OpenAI servers keeps going..

**Comment by jridgewell** [comment]
Hi @enricoros: I'm not sure where your code is coming from, can you provide a link? Just based on reading the screenshot, it looks like **you** are the keeping the connection alive by eagerly pulling the data out of `upstreamResponse.body` with a `for await (…)` loop (this is talked about in the [back-pressure and cancellation](https://sdk.vercel.ai/docs/concepts/backpressure-and-cancellation) doc).

I'm assuming your `eventParser` is a SSE parser, but I'm not sure where you're sending the data after it's parsed. It's possible this could be fixed with by switching to a `TransformStream` similar to https://github.com/vercel-labs/ai/blob/107e436e925f660ea9fd02ced726a02cb7831a25/packages/core/streams/ai-stream.ts#L31-L62

**Comment by enricoros** [comment]
Ok I will try it and report.
Many of us GPT UIs had implementations that predate vercel/ai, and possibly with a code path that's non optimal.

Code below:

https://github.com/enricoros/big-agi/blob/main/pages/api/openai/stream-chat.ts#L112

**Comment by jridgewell** [comment]
Reading your code, it's definitely possible to switch to a `TrasnformStream`:
- Move (most) of the code out of `start` handler, you can do it immediately before constructing the `TransformStream`
- Delete the `for await (…)` code, replace it with a `transform` handler as explained in the [back-pressure and cancellation](https://sdk.vercel.ai/docs/concepts/backpressure-and-cancellation)

**Comment by jvandenaardweg** [comment]
To hook into this conversation. My implementation looks like this: https://sdk.vercel.ai/docs/api-reference/openai-stream . So with all the methods the AI SDK provides. And the latest NextJS canary.

I also tried passing in `req.signal` into the `createChatCompletion` options, as it is an option there. But it does not seem to help. The server just throws the "Error: aborted" as mentioned by @enricoros , and then that's it. Also no more logging. But OpenAI is still sending data until it's done, i'm just not receiving it anymore.

**Comment by jridgewell** [comment]
@jvandenaardweg: It seems the `AbortSignal` isn't actually hooked up the node response that Next.js receives. I've opened https://github.com/vercel/next.js/pull/51727 to address.

And, I've discovered that the `cancel` handler in a `ReadableStream`/`TransformStream` isn't called either. That'll be fixed once https://github.com/vercel/edge-runtime/pull/428 is merged and we update the `edge-runtime` internal dependency.

**Comment by jvandenaardweg** [comment]
@jridgewell awesome! Appreciate the quick response on this! Looking forward to try it out 👍

**Comment by StringKe** [comment]
@jridgewell hi
Your commit solves this problem at the same time https://github.com/vercel/next.js/issues/50804 

I did not detect the same problem in version v13.4.8-canary.5.

**Comment by enricoros** [comment]
> Reading your code, it's definitely possible to switch to a `TrasnformStream`:
> 
> * Move (most) of the code out of `start` handler, you can do it immediately before constructing the `TransformStream`
> * Delete the `for await (…)` code, replace it with a `transform` handler as explained in the [back-pressure and cancellation](https://sdk.vercel.ai/docs/concepts/backpressure-and-cancellation)

Thanks for your help @jridgewell . Our app is now ported to use backpressure and cancellation, as you suggested. https://github.com/enricoros/big-agi/blob/490f8bdac30267662bee6b853ec8a3a303d2ab13/pages/api/llms/stream.ts#L141

I looked at the `vercel-labs/ai` implementation and adapted ours. Due to a couple of changes to the transformation functions, I couldn't use `vercel-labs/ai` as-is, but it's been an enormous help - thanks!

**Comment by jridgewell** [comment]
[13.4.8](https://www.npmjs.com/package/next/v/13.4.8) is out now, which fixes both issues from https://github.com/vercel-labs/ai/issues/90#issuecomment-1605084732.



> * with 13.4.8-canary.8, the TransformStream stops, BUT data is still sent from the node process to the OpenAI servers

The fetch between OpenAI and your server maintains its own buffering, so this is expected. When using the `AbortSignal` (or `cancel()` handler on the `ReadableStream`), you should be able to detect the client disconnect and abort the fetch to OpenAI.

**Comment by jensen** [comment]
Thank you for all of your work on this @jridgewell. This closes a long-standing support ticket I opened in April and allows me to give my client some options.

**Comment by enricoros** [comment]
@jridgewell: tested with `13.4.8` and it works!

In our [implementation](https://github.com/enricoros/big-agi/blob/main/pages/api/llms/stream.ts) (which is inspired by [AIStream](https://github.com/vercel-labs/ai/blob/main/packages/core/streams/ai-stream.ts), using a TransformStream) I can see the **connection stopping from the Node process to the OpenAI servers**! GREAT!

There's still an error message on the console (- error uncaughtException: Error: aborted), and maybe others won't see that, but apart from the scare effect, all the new changes seem to be working well! Our abort on the (browser) client side will stop the TransformStream on the (edge) server side, and the fetch to the OpenAI servers stops transmitting bytes too!

Well done @jridgewell!

**Comment by jvandenaardweg** [comment]
Thanks @jridgewell , confirmed it works! Token usages reported on the OpenAI website matches with what you would expect when you cancel the stream.

One future improvement could be to catch the abort in the Edge Function so we don't have a uncaught error, and allow to handle the abort, if that's even possible. In my use case I need to keep track of how many tokens are used. I already do this when started (`onStart`) for the prompt and when completed (`onCompletion`) for the generated output tokens. So handling the abort would allow me to report token usage at the moment it was aborted. But I think that would fit better in a new issue here on Github.

Many thanks all!

**Comment by enricoros** [comment]
> One future improvement could be to catch the abort in the Edge Function so we don't have a uncaught error, and allow to handle the abort

We're on the same use cases @jvandenaardweg 👍 good request!!

**Comment by wsxiaoys** [comment]
If you found `abort` signal is not fired on node 16, it's being discussed in https://github.com/vercel/next.js/pull/51727#issuecomment-1627745360

**Comment by jridgewell** [comment]
I'm working on Node 16 support in https://github.com/vercel/next.js/pull/52281, it's currently blocked on a test that only fails in CI.

**Comment by lassegit** [comment]
@jvandenaardweg Same here. Any resolution yet?

**Comment by jvandenaardweg** [comment]
> @jvandenaardweg Same here. Any resolution yet?

Unfortunately no. Just 2 possible workarounds I can think of:

1. When a user initiates a cancel from your UI, send the AI generated text to an API that tracks the token/character/word usage of that text. Probable downside is when a user navigates away from your site (or closes the browser), then the API call will not be fired.

2. Track each generated token using `onToken` on the server. And after a certain timeout just assume its cancelled. Also track the completion, so you can cancel them out. But for this you need to keep a record in a Redis store or something. Send each token to that store. Because once the stream is done, the Edge Function is terminated and your timeout code will not run. Then I guess you need to poll Redis each minute or so and pull that usage info into your main database.

I currently have option 1 in place in my app. Not optimal, but in terms of dev time the easiest I think.

**Comment by OlegGulevskyy** [comment]
Hm, how did you guys manage to make `stop` function working? What it does for me is cancels the stream but then it resumes it again by itself, therefore it never really does cancel it.
Nextjs version is `^14.1.0` and ai package version is `^2.2.31`
I would appreciate any pointers if there is something else needs to be implemented on the server side, but the [plain example](https://sdk.vercel.ai/docs/api-reference/use-completion) does not seem to work

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
The issue documents a concrete failure of the SDK's documented streaming cancellation behavior in the Edge/Next.js runtime. Although the root cause and fixes were in runtime dependencies, this was broken existing behavior rather than a request for a new capability.

### Evidence
- AbortController cancellation stopped client display but initially allowed OpenAI generation and token usage to continue in the background.
- A maintainer confirmed that Next.js had not connected the AbortSignal to its response and that stream cancel handlers were not invoked.
- Next.js 13.4.8 shipped fixes, after which multiple commenters confirmed that upstream OpenAI requests and token usage stopped as expected.
